src/kernel/serial.zig

```diff
@@ -1,80 +1,197 @@
 const arch = @import("arch.zig").internals;
+const panic = @import("panic.zig").panic;
+const testing = @import("std").testing;
+const options = @import("build_options");
 
-// The LCR is the line control register
-const LCR = 3;
-// Maximum baudrate
-const BAUD_MAX = 115200;
-// 8 bits per serial character
-const CHAR_LEN = 8;
-// One stop bit per transmission
-const SINGLE_STOP_BIT = true;
-// No parity bit
-const PARITY_BIT = false;
-
-pub const DEFAULT_BAUDRATE = 38400;
-
-const SerialError = error{InvalidBaud};
-
+/// The I/O port numbers associated with each serial port
 pub const Port = enum(u16) {
```

Comment:
Using an enum means it's impossible to send an invalid port number to the serial functions. E.g:

```Zig
const Port = enum(u16) { A = 123, B = 456, };
fn send(port: Port, byte: u8) void { }
send(999, 0); // Impossible
```
```Zig
fn send(port: u16, byte: u8) void {}
send(999, 0); // possible
```

Reply:
You could have 4 functions to send data to one of the COM ports. So you don't have to provide the port

Reply:
Nooooooo, I don't like the sound of that xD

Reply:
IMO it's better to have modularity by providing a function with a parameter rather than 4 functions with hardcode parameters, It makes adding more ports easier too.

Reply:
Another way is to have the port comptime and assert that the port is one of the 4 values. So if someone used the wrong port, it will produce a compile error.

Reply:
Sure that would work, but it would be overly restrictive as you wouldn't be able to decide which port to write to at runtime.

Reply:
It's still usable at runtime, its just that the port is known at compile time. And as there are only 4 options, there should be any restrictions. I use this method for my PIT.

Reply:
It would usable at runtime, but you couldn't write to a port that is only known at runtime. Consider a program that writes to a user-defined port:

```Zig
const serial = @import("serial.zig");
pub fn main() void {
    const cli_args = ...;
    const port = toInt(cli_args[0]);
    const byte = toInt(cli_args[1]);
    serial.write(port, byte); // compile error, port isn't known at compile time
}
```

You couldn't do that, and I see no reason why that should be made impossible. I don't see the point of using an integer over an enum value as an enum is more robust.

Reply:
I see. But in the example above, you will have to convert to a enum by switching on the int. So can have the port as an int and the `serial.write()` can return an error if the port is invalid.

Reply:
For converting the int to the enum you would just use `@intToEnum`. Isn't it better to make it impossible to provide an invalid port rather than make it possible and check? The outcome is the same but there are fewer steps and places to go wrong in between.